rfc/egress-billing.md

```diff
@@ -0,0 +1,309 @@
+  // Subject is storage node because they own that resource
+  "sub": "did:web:asia.web3.storage",
+  "pol": [
+    // Request origin header must be example.com
+    ["==", ".headers['origin']", "example.com"],
+    ["==", ".query.token", "zrptvx"]
+  ]
+}
+```
+
+For any given CID, the gateway will be able to look up the CID-specific content commitment as well as the
+group-to-gateway content commitment and combine them to create a chain rooted in the original content commitment
+provided by the w3up service on content upload. The gateway should take care to check that each link in the 
+chain is valid (though this will hopefully be handled by the UCAN tooling).
+
+```mermaid
+sequenceDiagram
+  Alice->>Blob Service: Please store bafk...7fi
+  Blob Service-->>Alice: Stored - you can present bafy...site to fetch bafk...7fi
+  Alice-)Group: You can present bafy...group to fetch bafk...7fi
+  Group-)Gateway: You can present bafy..access to fetch bafk...7fi
+```
+
+Note that `bafy..access` alone will not be enough to authorize retrieval
+of the content - the accounting system will need to combine it with
+a CID-specific content commitment that has been delegated to `Group`.
+
+## Implementation Considerations
+
+As stated above, a core design goal is to keep reads extremely performant. This generally implies
+that we should try to make a decision about whether or not to serve a request with as little IO (and to
+a lesser extent compute) as possible. As we suggest above, a service like Redis is designed to serve
+this use-case (https://redis.io/glossary/rate-limiting/) well, and ideally will be the only service
+consulted when making a decision about whether or not to serve content.
+
+This implies that the first time we are asked to serve a piece of content, we will serve it. This is
+totally acceptable because we propose a global rate limit greater than 0. After serving the content, we will
+further process the request using a background job-execution framework. This background job will determine 
+whether a particular CID should be rate limited. In some cases, especially cases where the same CID is being
+requested many times, this may result in content being served despite being over all applicable rate limits - this
+is acceptable since Redis will eventually be updated and the amount of reads that fall into this category are
+expected to be within the bounds of the read capacity the business is willing to serve.
+
+On the topic of Redis, it's worth noting that it is no longer open source:
+
+https://redis.io/blog/redis-adopts-dual-source-available-licensing/
+
+We may want to consider using an open source alternative like https://github.com/nalgeon/redka if we
+are concerned about using "source available" products - my inclination is to use Redis but consider products like
+Redka as a fallback in case Redis no longer meets our needs. We may also want to consider using AWS or Cloudflare products with Redis-compatible APIs:
```

Comment:
Cloudflare's KV is _super_ fast since it's eventually consistent, which feels like a good fit.

Reply:
ah cool will add it to the list!

I will say that one of the core requirements here is the ability to implement efficient rate limits - you can see an example of what that could mean here: https://redis.io/glossary/rate-limiting/

I think (but am not sure) that one of the key things is support for the `INCR` operation, which I don't think KV implements, and expiring entries, which may be better supported? 

The good news is that lots of stuff seems to implement the Redis query interface these days, including Cloudflare's Upstash and no less than two different AWS products - I think we'll have options!